File: test/language/array_types/python/PackedAutoArrayStructRecursionTest.py
import unittest
import zserio

from testutils import getZserioApi

class PackedAutoArrayStructRecursionTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "array_types.zs").packed_auto_array_struct_recursion

    def testBitSizeOfLength1(self):
        self._checkBitSizeOf(self.AUTO_ARRAY_LENGTH1)

    def testBitSizeOfLength2(self):
        self._checkBitSizeOf(self.AUTO_ARRAY_LENGTH2)

    def testBitSizeOfLength3(self):
        self._checkBitSizeOf(self.AUTO_ARRAY_LENGTH3)

    def testInitializeOffsetsLength1(self):
        self._checkInitializeOffsets(self.AUTO_ARRAY_LENGTH1)

    def testInitializeOffsetsLength2(self):
        self._checkInitializeOffsets(self.AUTO_ARRAY_LENGTH2)

    def testInitializeOffsetsLength3(self):
        self._checkInitializeOffsets(self.AUTO_ARRAY_LENGTH3)

    def testReadLength1(self):
        self._checkRead(self.AUTO_ARRAY_LENGTH1)

    def testReadLength2(self):
        self._checkRead(self.AUTO_ARRAY_LENGTH2)

    def testReadLength3(self):
        self._checkRead(self.AUTO_ARRAY_LENGTH3)

    def testWriteLength1(self):
        self._checkWrite(self.AUTO_ARRAY_LENGTH1)

    def testWriteLength2(self):
        self._checkWrite(self.AUTO_ARRAY_LENGTH2)

    def testWriteLength3(self):
        self._checkWrite(self.AUTO_ARRAY_LENGTH3)

    def _checkBitSizeOf(self, numElements):
        packedAutoArrayRecursion = self._createPackedAutoArrayRecursion(numElements)
        bitPosition = 2
        autoArrayRecursionBitSize = 8 + 8 + 1 + 6 + 8 + 8 + (numElements - 1) * (8 + 2)
        self.assertEqual(autoArrayRecursionBitSize, packedAutoArrayRecursion.bitsizeof(bitPosition))

    def _checkInitializeOffsets(self, numElements):
        packedAutoArrayRecursion = self._createPackedAutoArrayRecursion(numElements)
        bitPosition = 2
        expectedEndBitPosition = bitPosition + 8 + 8 + 1 + 6 + 8 + 8 + (numElements - 1) * (8 + 2)
        self.assertEqual(expectedEndBitPosition, packedAutoArrayRecursion.initialize_offsets(bitPosition))

    def _checkRead(self, numElements):
        writer = zserio.BitStreamWriter()
        self._writePackedAutoArrayRecursionToStream(writer, numElements)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        packedAutoArrayRecursion = self.api.PackedAutoArrayRecursion.from_reader(reader)
        self._checkPackedAutoArrayRecursion(packedAutoArrayRecursion, numElements)

    def _checkWrite(self, numElements):
        packedAutoArrayRecursion = self._createPackedAutoArrayRecursion(numElements)
        bitBuffer = zserio.serialize(packedAutoArrayRecursion)
        readPackedAutoArrayRecursion = zserio.deserialize(self.api.PackedAutoArrayRecursion, bitBuffer)
        self._checkPackedAutoArrayRecursion(readPackedAutoArrayRecursion, numElements)

    def _createPackedAutoArrayRecursion(self, numElements):
        autoArray = []
        for i in range(1, numElements + 1):
            element = self.api.PackedAutoArrayRecursion(i, [])
            autoArray.append(element)

        return self.api.PackedAutoArrayRecursion(0, autoArray)

    def _checkPackedAutoArrayRecursion(self, packedAutoArrayRecursion, numElements):
        self.assertEqual(0, packedAutoArrayRecursion.id)
        autoArray = packedAutoArrayRecursion.packed_auto_array_recursion
        self.assertEqual(numElements, len(autoArray))
        for i in range(1, numElements + 1):
            element = autoArray[i - 1]
            self.assertEqual(i, element.id)
            self.assertEqual(0, len(element.packed_auto_array_recursion))

    @staticmethod
    def _writePackedAutoArrayRecursionToStream(writer, numElements):
        writer.write_bits(0, 8)
        writer.write_varsize(numElements)
        writer.write_bool(True)
        maxBitNumber = 1
        writer.write_bits(maxBitNumber, 6)
        writer.write_bits(1, 8)
        writer.write_varsize(0)
        for _ in range(numElements - 1):
            writer.write_signed_bits(1, maxBitNumber + 1)
            writer.write_varsize(0)

    AUTO_ARRAY_LENGTH1 = 1
    AUTO_ARRAY_LENGTH2 = 5
    AUTO_ARRAY_LENGTH3 = 10

File: test/language/array_types/cpp/PackedAutoArrayStructRecursionTest.cpp
#include "gtest/gtest.h"

#include "array_types/packed_auto_array_struct_recursion/PackedAutoArrayRecursion.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

namespace array_types
{
namespace packed_auto_array_struct_recursion
{

class PackedAutoArrayStructRecursionTest : public ::testing::Test
{
protected:
    void fillPackedAutoArrayRecursion(PackedAutoArrayRecursion& packedAutoArrayRecursion, size_t numElements)
    {
        packedAutoArrayRecursion.setId(0);

        auto& autoArray = packedAutoArrayRecursion.getPackedAutoArrayRecursion();
        for (size_t i = 1; i <= numElements; ++i)
        {
            PackedAutoArrayRecursion element;
            element.setId(static_cast<uint8_t>(i));
            autoArray.push_back(element);
        }
    }

    size_t getPackedAutoArrayRecursionBitSize(size_t numElements)
    {
        size_t bitSize = 8; // id
        bitSize += 8; // varsize (length of auto array)
        bitSize += 1; // packing descriptor: isPacked
        if (numElements > 1)
            bitSize += 6; // packing descriptor: maxBitNumber
        bitSize += 8 + 8; // first element
        bitSize += (numElements - 1) * (8 + 2); // all deltas

        return bitSize;
    }

    void writePackedAutoArrayRecursionToByteArray(zserio::BitStreamWriter& writer, size_t numElements)
    {
        writer.writeBits(0, 8);
        writer.writeVarSize(static_cast<uint32_t>(numElements));
        writer.writeBool(true);
        const uint8_t maxBitNumber = 1;
        writer.writeBits(maxBitNumber, 6);
        writer.writeBits(1, 8);
        writer.writeVarSize(0);
        for (size_t i = 1; i <= numElements; ++i)
        {
            writer.writeSignedBits(1, maxBitNumber + 1);
            writer.writeVarSize(0);
        }
    }

    void checkPackedAutoArrayRecursion(const PackedAutoArrayRecursion& packedAutoArrayRecursion,
            size_t numElements)
    {
        ASSERT_EQ(0, packedAutoArrayRecursion.getId());
        const auto& autoArray = packedAutoArrayRecursion.getPackedAutoArrayRecursion();
        ASSERT_EQ(numElements, autoArray.size());
        for (size_t i = 1; i <= numElements; ++i)
        {
            const PackedAutoArrayRecursion& element = autoArray.at(i - 1);
            ASSERT_EQ(i, element.getId());
            ASSERT_EQ(0, element.getPackedAutoArrayRecursion().size());
        }
    }

    void checkBitSizeOf(size_t numElements)
    {
        PackedAutoArrayRecursion packedAutoArrayRecursion;
        fillPackedAutoArrayRecursion(packedAutoArrayRecursion, numElements);

        const size_t bitPosition = 2;
        const size_t autoArrayBitSize = getPackedAutoArrayRecursionBitSize(numElements);
        ASSERT_EQ(autoArrayBitSize, packedAutoArrayRecursion.bitSizeOf(bitPosition));
    }

    void checkInitializeOffsets(size_t numElements)
    {
        PackedAutoArrayRecursion packedAutoArrayRecursion;
        fillPackedAutoArrayRecursion(packedAutoArrayRecursion, numElements);

        const size_t bitPosition = 2;
        const size_t expectedEndBitPosition = bitPosition + getPackedAutoArrayRecursionBitSize(numElements);
        ASSERT_EQ(expectedEndBitPosition, packedAutoArrayRecursion.initializeOffsets(bitPosition));
    }

    void checkReadConstructor(size_t numElements)
    {
        zserio::BitStreamWriter writer(bitBuffer);
        writePackedAutoArrayRecursionToByteArray(writer, numElements);

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        PackedAutoArrayRecursion packedAutoArrayRecursion(reader);
        checkPackedAutoArrayRecursion(packedAutoArrayRecursion, numElements);
    }

    void checkWrite(size_t numElements)
    {
        PackedAutoArrayRecursion packedAutoArrayRecursion;
        fillPackedAutoArrayRecursion(packedAutoArrayRecursion, numElements);

        zserio::BitStreamWriter writer(bitBuffer);
        packedAutoArrayRecursion.write(writer);

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        PackedAutoArrayRecursion readPackedAutoArrayRecursion(reader);
        checkPackedAutoArrayRecursion(readPackedAutoArrayRecursion, numElements);
    }

    static const size_t AUTO_ARRAY_LENGTH1;
    static const size_t AUTO_ARRAY_LENGTH2;
    static const size_t AUTO_ARRAY_LENGTH3;
    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const size_t PackedAutoArrayStructRecursionTest::AUTO_ARRAY_LENGTH1 = 1;
const size_t PackedAutoArrayStructRecursionTest::AUTO_ARRAY_LENGTH2 = 5;
const size_t PackedAutoArrayStructRecursionTest::AUTO_ARRAY_LENGTH3 = 10;

TEST_F(PackedAutoArrayStructRecursionTest, bitSizeOfLength1)
{
    checkBitSizeOf(AUTO_ARRAY_LENGTH1);
}

TEST_F(PackedAutoArrayStructRecursionTest, bitSizeOfLength2)
{
    checkBitSizeOf(AUTO_ARRAY_LENGTH2);
}

TEST_F(PackedAutoArrayStructRecursionTest, bitSizeOfLength3)
{
    checkBitSizeOf(AUTO_ARRAY_LENGTH2);
}

TEST_F(PackedAutoArrayStructRecursionTest, initializeOffsetsLength1)
{
    checkInitializeOffsets(AUTO_ARRAY_LENGTH1);
}

TEST_F(PackedAutoArrayStructRecursionTest, initializeOffsetsLength2)
{
    checkInitializeOffsets(AUTO_ARRAY_LENGTH2);
}

TEST_F(PackedAutoArrayStructRecursionTest, initializeOffsetsLength3)
{
    checkInitializeOffsets(AUTO_ARRAY_LENGTH3);
}

TEST_F(PackedAutoArrayStructRecursionTest, readConstructorLength1)
{
    checkReadConstructor(AUTO_ARRAY_LENGTH1);
}

TEST_F(PackedAutoArrayStructRecursionTest, readConstructorLength2)
{
    checkReadConstructor(AUTO_ARRAY_LENGTH2);
}

TEST_F(PackedAutoArrayStructRecursionTest, readConstructorLength3)
{
    checkReadConstructor(AUTO_ARRAY_LENGTH3);
}

TEST_F(PackedAutoArrayStructRecursionTest, writeLength1)
{
    checkWrite(AUTO_ARRAY_LENGTH1);
}

TEST_F(PackedAutoArrayStructRecursionTest, writeLength2)
{
    checkWrite(AUTO_ARRAY_LENGTH2);
}

TEST_F(PackedAutoArrayStructRecursionTest, writeLength3)
{
    checkWrite(AUTO_ARRAY_LENGTH3);
}

} // namespace packed_auto_array_struct_recursion
} // namespace array_types
